pkg/planner/core/indexmerge_unfinished_path.go

```diff
@@ -0,0 +1,416 @@
+       [nil, unfinishedAccessPath{idx2,2 member of j->'$.b'}]
+     ]
+  }
+*/
+func generateUnfinishedIndexMergePathFromORList(
+	ds *DataSource,
+	orList []expression.Expression,
+	candidateAccessPaths []*util.AccessPath,
+) *unfinishedAccessPath {
+	if len(orList) < 2 {
+		return nil
+	}
+	unfinishedPartialPaths := make([]unfinishedAccessPathList, 0, len(orList))
+	for _, singleFilter := range orList {
+		unfinishedPathList := initUnfinishedPathsFromExpr(ds, candidateAccessPaths, singleFilter)
+		if unfinishedPathList == nil {
+			return nil
+		}
+		unfinishedPartialPaths = append(unfinishedPartialPaths, unfinishedPathList)
+	}
+	return &unfinishedAccessPath{
+		indexMergeORPartialPaths: unfinishedPartialPaths,
+	}
+}
+
+// initUnfinishedPathsFromExpr tries to collect access filters from the input filter for each candidate access path,
+// and returns them as a slice of unfinishedAccessPath, each of which corresponds to an input candidate access path.
+// If we failed to collect access filters for one candidate access path, the corresponding element in the return slice
+// will be nil.
+// If we failed to collect access filters for all candidate access paths, this function will return nil.
+/*
+Example1 (consistent with the one in generateUnfinishedIndexMergePathFromORList()):
+  Input:
+    expr: 1 member of j->'$.a'
+    candidateAccessPaths: [idx1(a, j->'$.a' unsigned array), idx2(j->'$.b' unsigned array, a)]
+  Output:
+    [unfinishedAccessPath{idx1,1 member of j->'$.a'}, nil]
+
+Example2:
+  Input:
+    expr: a = 3
+    candidateAccessPaths: [idx1(a, j->'$.a' unsigned array), idx2(j->'$.b' unsigned array, a)]
+  Output:
+    [unfinishedAccessPath{idx1,a=3}, unfinishedAccessPath{idx2,a=3}]
+*/
+func initUnfinishedPathsFromExpr(
+	ds *DataSource,
+	candidateAccessPaths []*util.AccessPath,
+	expr expression.Expression,
+) unfinishedAccessPathList {
+	retValues := make([]unfinishedAccessPath, len(candidateAccessPaths))
+	ret := make([]*unfinishedAccessPath, 0, len(candidateAccessPaths))
+	for i := range candidateAccessPaths {
+		ret = append(ret, &retValues[i])
+	}
+	for i, path := range candidateAccessPaths {
+		ret[i].index = path.Index
+		// case 1: try to use the previous logic to handle non-mv index
+		if !isMVIndexPath(path) {
+			// generateNormalIndexPartialPaths4DNF is introduced for handle a slice of DNF items and a slice of
+			// candidate AccessPaths before, now we reuse it to handle single filter and single candidate AccessPath,
+			// so we need to wrap them in a slice here.
+			paths, needSelection, usedMap := ds.generateNormalIndexPartialPaths4DNF(
+				[]expression.Expression{expr},
+				[]*util.AccessPath{path},
+			)
+			if len(usedMap) == 1 && usedMap[0] && len(paths) == 1 {
+				ret[i].initedAsFinished = true
+				ret[i].accessFilters = paths[0].AccessConds
+				ret[i].needKeepFilter = needSelection
+				continue
+			}
+		}
+		if path.IsTablePath() {
+			continue
```

Comment:
why pk is not considered here?

Reply:
Because I don't have enough time to make `PrepareIdxColsAndUnwrapArrayType()` support an access path without a `*model.IndexInfo`.
This limitation should not break previous behaviors because they should have been handled in the "case 1" above.

Reply:
soga, get

Reply:
soga, get